core/tests/test_apps.bitcoin.approver.py

```diff
@@ -0,0 +1,168 @@
+from common import unittest, await_result, H_
+
+from trezor import wire
+from trezor.messages.AuthorizeCoinJoin import AuthorizeCoinJoin
+from trezor.messages.TxInputType import TxInputType
+from trezor.messages.TxOutputType import TxOutputType
+from trezor.messages.SignTx import SignTx
+from trezor.messages import InputScriptType, OutputScriptType
+
+from apps.common import coins
+from apps.bitcoin.authorization import CoinJoinAuthorization
+from apps.bitcoin.sign_tx.approvers import CoinJoinApprover
+
+
+class TestApprover(unittest.TestCase):
+
+    def test_coinjoin_lots_of_inputs(self):
+        coin = coins.by_name('Bitcoin')
+        fee_per_anonymity_percent = 0.003
+        denomination = 10000000
+
+        # Other's inputs.
+        inputs = [
+            TxInputType(
+                amount=denomination + 1000000 * (i + 1),
+                script_type=InputScriptType.EXTERNAL,
+            ) for i in range(99)
+        ]
+
+        # Our input.
+        inputs.insert(
+            30,
+            TxInputType(
+                address_n=[H_(84), H_(0), H_(0), 0, 1],
+                amount=denomination + 1000000,
+                script_type=InputScriptType.SPENDWITNESS,
+            )
+        )
+
+        # Other's CoinJoined outputs.
+        outputs = [
+            TxOutputType(
+                amount=denomination,
+                script_type=OutputScriptType.PAYTOWITNESS,
+            ) for i in range(99)
+        ]
+
+        # Our CoinJoined output.
+        outputs.insert(
+            40,
+            TxOutputType(
+                address_n=[H_(84), H_(0), H_(0), 0, 2],
+                amount=denomination,
+                script_type=OutputScriptType.PAYTOWITNESS,
+            )
+        )
+
+        coordinator_fee = fee_per_anonymity_percent / 100 * len(outputs) * denomination
+        fees = coordinator_fee + 10000
+        total_coordinator_fee = coordinator_fee * len(outputs)
+
+        # Other's change-outputs.
+        outputs.extend(
+            TxOutputType(
+                amount=1000000 * (i + 1) - fees,
+                script_type=OutputScriptType.PAYTOWITNESS,
+            ) for i in range(99)
+        )
+
+        # Our change-output.
+        outputs.append(
+            TxOutputType(
+                address_n=[H_(84), H_(0), H_(0), 1, 1],
+                amount=1000000 - fees,
+                script_type=OutputScriptType.PAYTOWITNESS,
+            )
+        )
+
+        # Coordinator's output.
+        outputs.append(
+            TxOutputType(
+                amount=total_coordinator_fee,
+                script_type=OutputScriptType.PAYTOWITNESS,
+            )
+        )
+
+        msg_auth = AuthorizeCoinJoin(
```

Comment:
perhaps worth moving to `def setUp` too, if the authorization parameters are always the same?

Reply:
https://github.com/trezor/trezor-firmware/pull/1127/commits/e9831759a8fc2912e47757487f7befb89f5d272e